matrix/src/armory/matrix/matrix.py

```diff
@@ -0,0 +1,267 @@
+        [{'a': 1, 'b': 3}, {'a': 1, 'b': 4}, {'a': 2, 'b': 3}]
+        >>> list(create_matrix()(a=[1, 2], b=lambda a: (3, 4) if a == 1 else (5, 6)))
+        [{'a': 1, 'b': 3}, {'a': 1, 'b': 4}, {'a': 2, 'b': 5}, {'a': 2, 'b': 6}]
+
+    Args:
+        worker_num: The index of this node among all nodes, used to determine
+            which partition of the matrix to return. If specified, `num_workers`
+            must also be provided.
+        num_workers: The count of all nodes, used to determine the size of the
+            partition of the matrix to return. If specified, `worker_num` must
+            also be provided.
+        prune: A callable to determine whether a row of the matrix should be
+            pruned. If the callable returns `True` for a given set of
+            parameters, that parameter set, or row, will be omitted from the
+            returned matrix.
+
+    Returns:
+        A generator of key-value mappings. The generator function accepts
+        keyword arguments of iterables, ranges, or sequences specifying the
+        allowable values for each argument. The yielded value of the generator
+        is a key-value mapping of arguments corresponding to a row of the
+        matrix.
+
+        If an argument to the generator is a callable, it will be called with
+        the preceding keyword parameter values for the current row. The return
+        of the callable is the value--or iterable--for the keyword, from which
+        to generate additional rows in the matrix.
+    """
+
+    if worker_num is None or num_workers is None:
+        worker_num = 0
+        num_workers = 1
+
+    def _generate(**kwargs) -> Iterable[Mapping[str, Any]]:
+        count = 0
+        for params in product({}, list(kwargs.items())):
+            # Skip over entries that are pruned
+            if prune is not None and prune(**params):
+                continue
+            # Skip over entries when partioning
+            if (count % num_workers) == worker_num:
+                yield params
+            count += 1
+
+    return _generate
+
+
+class Matrix:
+    """
+    A function to be invoked multiple times with parameters from rows of a
+    cartesian product matrix of all possible parameters.
+
+    This class is not intended to be used directly, but is used internally by
+    the `matrix` decorator.
+    """
+
+    def __init__(self, func: Callable, **kwargs):
+        self.func = func
+        self.kwargs = kwargs
+        self.worker_num: Optional[int] = None
+        self.num_workers: Optional[int] = None
+        self.pruner: Optional[Callable[..., bool]] = None
+
+    @property
+    def matrix(self):
+        """The generated matrix of arguments"""
+        return create_matrix(
+            worker_num=self.worker_num,
+            num_workers=self.num_workers,
+            prune=self.pruner,
+        )(**self.kwargs)
+
+    def __call__(self, *args, **kwargs):
```

Comment:
The `args` and `kwargs` for function arguments not specified as parameters should be mentioned in the README.